rmw_iceoryx_cpp/src/internal/iceoryx_topic_names_and_types.cpp

```diff
@@ -36,67 +38,78 @@ void fill_topic_containers(
   std::map<std::string, std::vector<std::string>> & subscribers_topics_,
   std::map<std::string, std::vector<std::string>> & publishers_topics_)
 {
-  static iox::popo::Subscriber port_receiver(iox::roudi::IntrospectionPortService);
+  static iox::popo::UntypedSubscriber port_receiver(iox::roudi::IntrospectionPortService);
   static std::map<std::string, std::string> names_n_types;
   static std::map<std::string, std::vector<std::string>> subscribers_topics;
   static std::map<std::string, std::vector<std::string>> publishers_topics;
 
   bool updated = false;
-  if (iox::popo::SubscriptionState::SUBSCRIBED != port_receiver.getSubscriptionState()) {
-    port_receiver.subscribe(1);
+  if (iox::SubscribeState::SUBSCRIBED != port_receiver.getSubscriptionState()) {
+    port_receiver.subscribe();
     // wait for delivery on subscribe
-    while (!port_receiver.hasNewChunks()) {
+    while (!port_receiver.hasData()) {
       std::this_thread::sleep_for(std::chrono::milliseconds(50));
     }
     updated = true;
   } else {
-    updated = port_receiver.hasNewChunks();
+    updated = port_receiver.hasData();
   }
 
   if (updated) {
     // get the latest sample
     const iox::mepoo::ChunkHeader * chunk_header = nullptr;
     const iox::mepoo::ChunkHeader * latest_chunk_header = nullptr;
 
-    while (port_receiver.getChunk(&chunk_header)) {
-      if (latest_chunk_header) {
-        port_receiver.releaseChunk(latest_chunk_header);
+    while(1)
+    {
+      auto result = port_receiver.take();
+      if (!result.has_error())
+      {
+        chunk_header = iox::mepoo::ChunkHeader::fromUserPayload(result.value());
+        if (latest_chunk_header)
+        {
+          port_receiver.release(latest_chunk_header);
+        }
+        latest_chunk_header = chunk_header;
+      }
+      else
+      {
+        break;
       }
-      latest_chunk_header = chunk_header;
     }
 
     if (latest_chunk_header) {
       const iox::roudi::PortIntrospectionFieldTopic * port_sample =
         static_cast<const iox::roudi::PortIntrospectionFieldTopic *>(latest_chunk_header->
-        payload());
+        userPayload());
```

Comment:
I think this could be done easier now.  With the old API we always returned the ChunkHeader. As you are interested in the payload only and a pointer to the payload is returned, you should be able to get rid of  `iox::mepoo::ChunkHeader::fromUserPayload` and  `userPayload()`. Just store and use the payload pointer. You could then also remove chunk_header and just use latest_chunk_header and update it if `!result.has_error()`.  port_introspection would then also be a better name. I even think you could use a typed Subscriber of type PortIntrospectionFieldTopic as you know exactly what you'll receive here

Reply:
Thanks, I will read the example code and refactor this.

Reply:
May this will be solved by one of your teammate mossmaurice